### PR TITLE
Draft: Randomizing executors in some unit tests

### PR DESCRIPTION
**Pull request by cbuescher**
In order to test concurrent search execution in unit tests more often, this draft shows some
places where we might want to randomly set ContextIndexSearcher executors or increase test document sizes to
increase the likelihood of creating more segments and thus more slices for concurrent search execution.

This PR deliberately modifies the concurrent search path to throw an IllegalStateException to get a feeling for how
likely this path is executed in tests.

### CODE REVIEW

File: server/src/test/java/org/elasticsearch/search/SearchCancellationTests.java [hunk 1]
```diff
@@ -79,14 +81,21 @@ public static void cleanup() throws IOException {
...
+        // TODO maybe randomize setting executor here
+        return new ContextIndexSearcher(
             reader,
             IndexSearcher.getDefaultSimilarity(),
             IndexSearcher.getDefaultQueryCache(),
             IndexSearcher.getDefaultQueryCachingPolicy(),
-            true
+            1,
+            true,
+            null
```
Comment (javanna) [inline server/src/test/java/org/elasticsearch/search/SearchCancellationTests.java:93]: I am thinking that we should have a newContextSearch method in ESTestCase, similar to LuceneTestCase#newSearcher , and take care of the randomization there. This centralizes tha randomized aspects, and also removes the need to fiddle with the executor in each test.

File: test/framework/src/main/java/org/elasticsearch/test/ESTestCase.java [hunk 2]
```diff
@@ -218,6 +224,7 @@ public abstract class ESTestCase extends LuceneTestCase {
     private static final Collection<String> loggedLeaks = new ArrayList<>();
 
     private HeaderWarningAppender headerWarningAppender;
+    private List<ThreadPoolExecutor> threadPoolExecutors = new ArrayList<>();
```
Comment (javanna) [inline test/framework/src/main/java/org/elasticsearch/test/ESTestCase.java:227]: I am thinking we could do something like we do now in Lucene, see https://github.com/apache/lucene/pull/12588 . We could create a single executor eagerly, and then reuse across all the searchers that are meant to get an executor.

File: test/framework/src/main/java/org/elasticsearch/test/ESTestCase.java [hunk 3]
```diff
@@ -1292,6 +1299,56 @@ public Environment newEnvironment(Settings settings) {
         return TestEnvironment.newEnvironment(build);
     }
 
+    /**
+     * Create a new {@link ContextIndexSearcher} from the given reader, randomizing if an executor for
+     * multi-threaded collection will be used or not
+     */
+    public ContextIndexSearcher newContextSearcher(IndexReader reader) throws IOException {
```
Comment (javanna) [inline test/framework/src/main/java/org/elasticsearch/test/ESTestCase.java:1306]: There is a newINdexSearcher method in AggregatorTestCase that seems like it's doing a very similar thing, maybe the two can be merged into one?

File: test/framework/src/main/java/org/elasticsearch/test/ESTestCase.java (hunk 2 above)
Comment (JVerwolf) [reply test/framework/src/main/java/org/elasticsearch/test/ESTestCase.java:227]: Thanks for the suggestion @javanna. I tried this and got some esoteric errors about thread leaks in some tests.  After a bunch of time hunting down the root cause, I realized that multiple executors can be created in a single test, and we need to keep track of all of them for cleanup purposes.

See for example `org.elasticsearch.xpack.core.security.authz.accesscontrol.SecurityIndexReaderWrapperIntegrationTests#testDLS` where ContextIndexSearcher is called in a `for` loop.

**Comment by JVerwolf** [comment]
I added https://github.com/elastic/elasticsearch/pull/100358 as a follow up to this investigation.

**Pull request closed without merging**